src/Microsoft.TemplateEngine.Edge/Settings/AsyncMutex.cs

```diff
@@ -0,0 +1,84 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#nullable enable
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.TemplateEngine.Edge.Settings
+{
+    /// <summary>
+    /// Helper class to work with <see cref="Mutex"/> in <c>async</c> method, since <c>await</c>
+    /// can switch to different thread and <see cref="Mutex.ReleaseMutex"/> must be called from same thread.
+    /// Hence this helper class.
+    /// </summary>
+    internal sealed class AsyncMutex : IDisposable
+    {
+        private readonly TaskCompletionSource<IDisposable> _taskCompletionSource;
+        private readonly ManualResetEvent _blockReleasingMutex = new ManualResetEvent(false);
+        private readonly string _mutexName;
+        private readonly CancellationToken _token;
+        private bool _disposed;
+
+        private AsyncMutex(string mutexName, CancellationToken token)
+        {
+            _mutexName = mutexName;
+            _token = token;
+            _taskCompletionSource = new TaskCompletionSource<IDisposable>();
+
+            var thread = new Thread(new ThreadStart(WaitLoop));
```

Comment:
Is it really necessary to create a new `Thread`? As opposed to using a thread pool.

Reply:
Switched to calling ThreadPool, at time it felt more natural to created dedicated thread for this, but after some more thinking I agree, switched to ThreadPool